xc/device.ml

```diff
@@ -1430,6 +1430,102 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
+      let wakeup () =  (* write single byte to wake up Monitor.wait *)
+        if Unix.write pipe_w " " 0 1 <> 1 then
+          debug "Pipe wrote error, failed to wake up qmp event thread"
+      let add m fd = Epoll.set m fd Flags.in_; wakeup ()
+      let remove m fd = Epoll.remove m fd; wakeup ()
+      let wait m = Epoll.wait m ~timeout:`Never
+      let with_event m fn = function
+        | `Ok -> Epoll.iter_ready m ~f:(fun fd flags -> fn fd (flags = Flags.in_))
+        | `Timeout -> debug "Shouldn't receive epoll timeout event in qmp_event_thread"
+    end
+    let m = Monitor.create ()
+
+    let monitor_path domid = Printf.sprintf "/var/run/xen/qmp-event-%d" domid
+    let debug_exn msg e = debug "%s: %s" msg (Printexc.to_string e)
+
+    let remove domid =
+      try
+        Lookup.channel_of domid >>= fun c ->
+        Lookup.remove c domid;
+        Monitor.remove m (Qmp_protocol.to_fd c);
+        Qmp_protocol.close c;
+        debug "Removed QMP Event fd for domain %d" domid
+      with e -> debug_exn (Printf.sprintf "Got exception trying remove qmp on domain-%d" domid) e
+
+    let add domid =
+      try
+        Lookup.channel_of domid >>| fun () ->
+        let c = Qmp_protocol.connect (monitor_path domid) in
+        Qmp_protocol.negotiate c;
+        Lookup.add c domid;
+        Monitor.add m (Qmp_protocol.to_fd c);
+        debug "Added QMP Event fd for domain %d" domid
+      with e ->
+        debug_exn (Printf.sprintf "QMP domain-%d: negotiation failed: removing socket" domid) e;
+        remove domid
+
+    let qmp_event_handle domid qmp_event =
+      (* This function will be extended to handle qmp events *)
+      debug "Got QMP event, domain-%d: %s" domid qmp_event.event
+
+    let qmp_event_thread () =
+      Monitor.add m pipe_r;
+      while true do
+        try
+          Monitor.wait m |> Monitor.with_event m (fun fd is_flag_in ->
+              match fd with
+              | pipe when pipe = pipe_r ->
+                if is_flag_in then ()
+                else debug "Received unexpected epoll event on pipe_r in qmp_event_thread"
+              | sock ->
+                Lookup.domid_of sock >>= fun domid ->
+                if is_flag_in then
+                  Lookup.channel_of domid >>= fun c ->
+                  try
+                    match Qmp_protocol.read c with
+                    | Event e -> qmp_event_handle domid e
+                    | msg -> debug "Got non-event message, domain-%d: %s" domid (string_of_message msg)
+                  with End_of_file ->
+                    debug "domain-%d: end of file, close qmp socket" domid;
+                    remove domid
+                else begin
+                  debug "EPOLL error on domain-%d, close qmp socket" domid;
+                  remove domid
+                end
+            )
+        with e ->
+          debug_exn "Exception in qmp_event_thread: %s" e;
+      done
+
+    let init_qmp_event =
```

Comment:
You can call it `let _init_qmp_event` to avoid the compiler complaining that the value is not used

Reply:
Updated, thanks.